packages/block-library/src/button/edit.js

```diff
@@ -133,10 +134,60 @@ function ButtonEdit( props ) {
 
 	const [ isEditingURL, setIsEditingURL ] = useState( false );
 	const isURLSet = !! url;
-	const opensInNewTab = linkTarget === NEW_TAB_TARGET;
-	const nofollow = !! rel?.includes( NOFOLLOW_REL );
+
 	const isLinkTag = 'a' === TagName;
 
+	// Defines how block attributes map to link value and vice versa.
+	const linkValueAttrsToDataMapping = {
+		url: 'url',
+		opensInNewTab: {
+			dataKey: 'linkTarget',
+			toLink: ( value ) => value === NEW_TAB_TARGET,
+			toData: ( value ) => ( value ? NEW_TAB_TARGET : undefined ),
+		},
+		nofollow: {
+			dataKey: 'rel',
+			toLink: ( value ) => value?.includes( NOFOLLOW_REL ),
+			toData: ( value, { opensInNewTab: opensInNewTabValue } ) => {
+				// "rel" attribute can be effected by changes to
+				// "nofollow" and "opensInNewTab" attributes.
+				// In addition it is editable in plaintext via the UI
+				// so consider that it may already contain a value.
+				let updatedRel = '';
+
+				// Handle setting rel based on nofollow setting.
+				if ( value ) {
+					updatedRel = updatedRel?.includes( NOFOLLOW_REL )
+						? updatedRel
+						: updatedRel + ` ${ NOFOLLOW_REL }`;
+				} else {
+					const relRegex = new RegExp(
+						`\\b${ NOFOLLOW_REL }\\s*`,
+						'g'
+					);
+					updatedRel = updatedRel?.replace( relRegex, '' ).trim();
+				}
+
+				// Handle setting rel based on opensInNewTab setting.
```

Comment:
What do we think about handling _another_ attribute in this handler? If we don't do it like this we will need some means within `toData` to trigger mutations in other parts of the data tree.

Whilst the approach here isn't one I love, it feels less complex to the consumer than any alternative I can conceive of right now.

Reply:
Also this might cause unexpected conditions because what if `opensInNewTab` were processed _after_ `nofollow`? That might mean that `opensInNewTab` value might still represent the _old_ value as it has yet to be processed.

We need a test for such a scenario to make it concrete and then we can consider how best to resolve it.

Reply:
Possibly a "dependencies" idea, whereby properties can declare that they have dependencies on other properties. Then the final value for the property is only calculated once the other property has been processed.

Reply:
@getdave I have thought about this, can we somehow define `util` methods for the mapping object
1. Function that returns mapping object
```
const mapping = ( () => {
    const utils = {
        method1: ()=>{},
        ...
    }
    return {
       'url': 'href'
       'nofollow': ()=>{ utils.method1(); } 
   }
} )();
```
2. Object with utilities methods
const mapping = {
    'url': 'href',
    'utils': {
        method1: ()=>{},
        ...
    }
}
With this approach, we may need to `exclude` the prop name from the `link-value-transforms` methods, since it does not have standard toLink and toData methods like normal props

Reply:
I made a PR for demo this https://github.com/WordPress/gutenberg/pull/55143

Reply:
I've also been thinking about ths concept of utility functions. I think on balance it might be easier to simply allow consumers to provide their own utilities as needed.

For example much of the transforms could be handled with a functional approach using something like `compose` to compose a set of transformations.

I think the most common one will be "if this value is a string then concat to an existing string if it exists or otherwise just use the string as the new value". Something like [Ramda Concat](https://ramdajs.com/docs/#concat) _but_ that can handle `undefined` would be what we need.

Reply:
Oh I've just realised the concept in your PR 👍 

So we could export an additional helper from `link-value-transforms` called `mapGenerator` where we could expose utils for mapping. 

If I've understood this would accept a mapping object and provide callbacks to the map to allow the consumer to perform typical data operations.

This would also be completely optional.

I think it's clever. I think we should 

- implement a basic version of _this_ (my) PR 
- come back to your idea of mapping generators in a follow up. 

I really like the idea but I think it may overcomplicate this initial PR. But I would definitely like to explore it.

What do you think?